modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/model/listener/CompanyModelListener.java

```diff
@@ -0,0 +1,37 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2024 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.frontend.data.set.admin.web.internal.model.listener;
+
+import com.liferay.frontend.data.set.admin.model.DataSetModelManager;
+import com.liferay.portal.kernel.exception.ModelListenerException;
+import com.liferay.portal.kernel.model.BaseModelListener;
+import com.liferay.portal.kernel.model.Company;
+import com.liferay.portal.kernel.model.ModelListener;
+import com.liferay.portal.kernel.transaction.TransactionCommitCallbackUtil;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Daniel Sanz
+ */
+@Component(service = ModelListener.class)
+public class CompanyModelListener extends BaseModelListener<Company> {
+
+	@Override
+	public void onAfterCreate(Company company) throws ModelListenerException {
+		TransactionCommitCallbackUtil.registerCallback(
```

Comment:
Does this ensure code inside executes in a transaction? If so, I would expect such code to be closer to `_generate` method. Something similar to `@Transactional` annotation or `extends BaseTransactionalMVCResourceCommand`.

Reply:
I used this because otherwise the admin user might not be created at the time of running this listener. It somehow makes sense because here we are listening for just the Company model, however we want portal instance to be completely created, which does not only entail the Company model but [quite a few things](https://github.com/liferay/liferay-portal/blob/57fb28672ac43550db25e43108400bf80f309eff/portal-impl/src/com/liferay/portal/service/impl/CompanyLocalServiceImpl.java#L235-L320).

This pattern is common in the CompanyModelListener classes around, when listener needs the transaction creating the company to finish

Reply:
Oh, that's what the util does. Seems a bit silly, a callback within a callback. It feels to me that there should be a `CompanyListener`, with `onAfterCreate` that does not require this weirdness. Who knows, maybe there is a good reason for this. In any case, this does look like the way to go for us :+1: